Clean up editors code

sergiou87 opened the pull request:
Closes #5409

## Description

The changes in this PR are an attempt to simplify the code that integrates the app with external editors. Main changes:
- Gather all the info relative to external editors into entries of an array. This should make it easier and clearer adding support for new editors, since now all that info and logic won't be scattered.
- Removed the `ExternalEditor` enum and use a string instead. Several reasons for this:
  1. No code was leveraging the advantages of `ExternalEditor` as an enum, other than in `editors/darwin.ts`, `editors/win32.ts` and `editors/linux.ts`. Everywhere else its value was passed around, and finally stored/read as a string in/from the app settings.
  1. There weren't references to specific values of the enum outside of the mentioned files.
  1. With the new list of editors, keeping the enum around was just redundant. In the end, only editors inside the list AND detected in the user's machine will be available and used (🐛s aside).
- macOS editors are now opened using `open`, which only needs us to provide the bundle identifier of the editor, so we don't need to keep hardcoding paths to specific executables inside the editors' bundle.

Remaining work in this PR:
- ~~Make the same changes in the `editors/linux.ts` file.~~
- ~~Update the `docs/technical/editor-integration.md` document to reflect these new changes.~~
- ~~Other cleanup!~~

## Release notes

Notes: no-notes

sergiou87 committed:
Clean up macOS editors code and use `open` instead of looking for the executable

sergiou87 committed:
Clean up Windows editors code? Hopefully!

sergiou87 committed:
Get rid of `ExternalEditor` since an enum it's not needed

sergiou87 committed:
Add some comments to the new editor entries

sergiou87 commented on app/src/lib/editors/shared.ts:
```diff
@@ -1,32 +1,11 @@
-import * as Win32 from './win32'
-import * as Linux from './linux'
-
-export type ExternalEditor = Darwin.ExternalEditor | Win32.ExternalEditor
```
Does removing `ExternalEditor` feel like we're losing _something valuable_?

As I mentioned in the PR description, it doesn't feel like we lost anything from a "compiler help" perspective. But maybe the code was clearer with it… maybe it's worth making `type ExternalEditor = string`? 🤔

sergiou87 commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,592 @@ import {
+   * @param displayName The display name as listed in the registry
+   * @param publisher The publisher who created the installer
+   */
+  readonly expectedInstallationChecker: ExpectedInstallationChecker
```
I'm not particularly proud of these properties or their names or anything, but I'm open to suggestions 😂 

This felt like the best way to avoid scattering specific editor stuff all over this file 🤔

niik replied:
I'm fine with these names but I would probably have called the `appInformationExtractor` method something like `getAppInfo` or `getUninstallInfo`. Your call tho

sergiou87 commented on app/src/lib/editors/win32.ts:
```diff
@@ -705,147 +637,28 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
+  const editorPaths = await Promise.all(
+    editors.map(editor =>
+      findApplication(editor).then(path => {
+        return { editor, path }
```
Ad-hoc type + TypeScript compiler help = 😍 

But does it "feel" right for a TypeScript coder? 😳

niik replied:
> But does it "feel" right for a TypeScript coder? 😳

Absolutely!

sergiou87 committed:
Improve comments 🤓

niik commented on app/src/lib/editors/darwin.ts:
```diff
@@ -1,174 +1,135 @@
       }
 
-      log.debug(`Command line interface for ${editor} not found at '${path}'`)
+      log.debug(`App instalation for ${editor} not found at '${installPath}'`)
```
```suggestion
      log.debug(`App installation for ${editor} not found at '${installPath}'`)
```

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -705,147 +640,28 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
-      path: webstormPath,
-    })
-  }
+  const editorPaths = await Promise.all(
```
I know that we used Promise.all in the previous implementation but I think we can make this method much more readable and compact by moving the call to `findApplication` inside the loop below. Since our registry-js calls are synchronous (unfortunately) it shouldn't alter the behavior in any noticeable way.

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -705,147 +640,28 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
+      results.push({
+        editor: editor.name,
+        path: path,
+        usesShell: editor.usesShell ?? false,
```
If we infer this from the path we can ditch  the `usesShell` from the `IFoundEditor` type.

```suggestion
        usesShell: path.endsWith('.cmd')
```

sergiou87 replied:
I thought about this, but I wasn't 100% sure this would always be the case for future editors we might support. Anyway, happy to make the change!

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -705,147 +640,28 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
+    if (path) {
+      results.push({
+        editor: editor.name,
+        path: path,
```
If the name of the local variable matches that of the field you're trying to assign to you can omit the right hand side.

```suggestion
        path,
```

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+      },
+    ],
+    executableShimPath: ['subl.exe'],
+    appInformationExtractor: keys => {
```
If we change the `expectedInstallationChecker` for this entry to check whether displayName starts with `Sublime Text` we should be able to remove this `appInformationExtractor` method

sergiou87 replied:
Nice catch!

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+      },
+    ],
+    executableShimPath: ['bin', 'webstorm.exe'],
+    appInformationExtractor: keys => {
```
Since the `expectedInstallationChecker` for this entry already checks whether displayName starts with `WebStorm` we should be able to remove this `appInformationExtractor` method, right?

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+      },
+    ],
+    executableShimPath: ['bin', 'phpstorm.exe'],
+    appInformationExtractor: keys => {
```
Similar to the `WebStorm` entry, this method looks to be redundant?

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+      },
+    ],
+    executableShimPath: [],
+    appInformationExtractor: keys => {
+      const displayName = getKeyOrEmpty(keys, 'DisplayName')
+      const publisher = getKeyOrEmpty(keys, 'Publisher')
+      const installLocation = getKeyOrEmpty(keys, 'DisplayIcon')
+
+      return { displayName, publisher, installLocation }
+    },
```
If we leverage the default extractor here we can make it clear that we're only overriding the `installLocation`
```suggestion
    appInformationExtractor: keys => ({
      ...defaultAppInformationExtractor(keys),
      installLocation: getKeyOrEmpty(keys, 'DisplayIcon'),
    }),
```

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+      },
+    ],
+    executableShimPath: ['bin', 'rider64.exe'],
+    appInformationExtractor: keys => {
```
Similar to `WebStorm` and `PhpStorm` we should be able to get rid of this method

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,663 +10,595 @@ import {
+  return { displayName, publisher, installLocation }
+}
 
+async function findApplication(
```
I know this isn't part of your refactor but upon reading this method again it sure feels weird that we're just grabbing the first set of keys and then not continuing down the fallback chain if the `expectedInstallationChecker`  fails.

Seems like this method should be rewritten such that we test the validity of each set of keys. Something like this:

```ts
async function findApplication(editor: IWindowsExternalEditor) {
  const getAppInfo =
    editor.appInformationExtractor ?? defaultAppInformationExtractor

  for (const { key, subKey } of editor.registryKeys) {
    const keys = enumerateValues(key, subKey)
    if (keys.length === 0) {
      continue
    }

    const { displayName, publisher, installLocation } = getAppInfo(keys)

    if (!editor.expectedInstallationChecker(displayName, publisher)) {
      log.debug(`Unexpectted registry entries for ${editor.name}`)
      continue
    }

    const path = Path.join(installLocation, ...editor.executableShimPath)
    const exists = await pathExists(path)
    if (!exists) {
      log.debug(`Executable for ${editor.name} not found at '${path}'`)
      continue
    }

    return path
  }

  return null
}
```

niik commented on app/src/lib/editors/darwin.ts:
```diff
@@ -182,150 +143,27 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
-    results.push({ editor: ExternalEditor.Nova, path: novaPath })
+  const results: Array<IFoundEditor<string>> = []
+
+  const editorPaths = await Promise.all(
```
Similar comment here as with the win32 `getAvailableEditors`. I think we should move the call to `findApplication` into the loop below.

niik commented on app/src/lib/editors/darwin.ts:
```diff
@@ -182,150 +143,27 @@ async function findApplication(editor: ExternalEditor): Promise<string | null> {
+    const { editor, path } = editorPath
+
+    if (path) {
+      results.push({
+        editor: editor.name,
+        path: path,
+      })
```
:art: 
```suggestion
      results.push({ editor: editor.name, path })
```

niik reviewed (changes requested):
This is such a fantastic improvement @sergiou87, thanks for working on this!

We'll also need to review and possibly update [our documentation](https://github.com/desktop/desktop/blob/development/docs/technical/editor-integration.md) for adding a new external editor.

I also think there's potential to reduce duplication in the win32 file by introducing a few helper functions to construct these registry keys. Something like:

```ts
const registryKey = (key: HKEY, ...subKeys: string[]) =>
  ({ key, subKey: Path.win32.join(...subKeys) })

const uninstallSubKey =
  'SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall'

const wow64UninstallSubKey =
  'SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall'

const CurrentUserUninstallKey = (subKey: string) =>
  registryKey(HKEY.HKEY_CURRENT_USER, uninstallSubKey, subKey)

const LocalMachineUninstallKey = (subKey: string) =>
  registryKey(HKEY.HKEY_LOCAL_MACHINE, uninstallSubKey, subKey)

const Wow64LocalMachineUninstallKey = (subKey: string) =>
  registryKey(HKEY.HKEY_LOCAL_MACHINE, wow64UninstallSubKey, subKey)
```

What do you think about that?

It'd cut down the VS code entry to just

```ts
    name: 'Visual Studio Code',
    registryKeys: [
      // 64-bit version of VSCode (user) - provided by default in 64-bit Windows
      CurrentUserUninstallKey('{771FD6B0-FA20-440A-A002-3B3BAC16DC50}_is1'),
      // 32-bit version of VSCode (user)
      CurrentUserUninstallKey('{D628A17A-9713-46BF-8D57-E671B46A741E}_is1'),
      // 64-bit version of VSCode (system) - was default before user scope installation
      LocalMachineUninstallKey('{EA457B21-F73E-494C-ACAB-524FDE069978}_is1'),
      // 32-bit version of VSCode (system)
      Wow64LocalMachineUninstallKey('{F8A2A208-72B3-4D61-95FC-8A65D340689B}_is1'),
    ],
```

sergiou87 committed:
Fix typos and rename things :sparkling_heart:

sergiou87 committed:
Delete more unnecessary code :heart_eyes:

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

sergiou87 committed:
More cleanup and a fix for how registry keys are iterated

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

sergiou87 committed:
MOAR cleanup :_D

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

sergiou87 committed:
Use some utility functions to clean up registry keys

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

sergiou87 commented:
Thank you for the thorough review and awesome feedback!! 🙇‍♂️ 🙇‍♂️ ❤️ 

> We'll also need to review and possibly update [our documentation](https://github.com/desktop/desktop/blob/development/docs/technical/editor-integration.md) for adding a new external editor.

Yup! That (and `linux.ts`) is in my todo list. I wanted to confirm this approach with you before updating the documentation 😄 

> I also think there's potential to reduce duplication in the win32 file by introducing a few helper functions to construct these registry keys. Something like:
> …

I absolutely love it 😍 

After your suggestions to remove many of the `appInformationExtractor` definitions, and given I still don't like its name 😂 , I realized we don't really need it anymore. Only Notepad++ has a different logic, and it's just it looks for the install location in a different registry key. So I just replaced the `appInformationExtractor` function with an optional `installLocationRegistryKey` that defaults to `'InstallLocation'` and is overridden for Notepad++ only.

I'll make the changes in `linux.ts`, in the docs, test everything again everywhere I can, and let you know when it can be re-reviewed 🤓

sergiou87 committed:
Update editor integration document to reflect Windows & macOS changes

sergiou87 committed:
Refactor editors code for Linux

sergiou87 committed:
Iterate through the paths in Linux editors 🤦‍♂️

sergiou87 committed:
More cleanup :_D

sergiou87 commented:
@niik I think I addressed all your feedback and also did some additional cleanup. Please, take a look at my comment above for more details 😌 

I also tested the few editors I have on Windows, Linux and macOS and everything worked as expected. On Linux I had to change locally a few executable paths though, since they didn't match what I had in my Ubuntu 🤷‍♂️

sergiou87 committed:
Merge remote-tracking branch 'origin/development' into external-editor-refactor

niik commented on app/src/lib/editors/darwin.ts:
```diff
@@ -1,174 +1,135 @@
+ * This list contains all the external editors supported on macOS. Add a new
+ * entry here to add support for your favorite editor.
+ **/
+const editors: IDarwinExternalEditor[] = [
```
Given how simple this type is you could simplify it to be a Record type keyed on the name if you wanted.  Something like

```ts
const editors: Record<string, ReadonlyArray<string>> = {
  Atom: ['com.github.atom'],
  MacVim: ['org.vim.MacVim'],
  'Visual Studio Code': ['com.microsoft.VSCode'],
  'Visual Studio Code (Insiders)': ['com.microsoft.VSCodeInsiders'],
  VSCodium: ['com.visualstudio.code.oss'],
  'Sublime Text': [
    'com.sublimetext.4',
    'com.sublimetext.3',
    'com.sublimetext.2',
  ],
  BBEdit: ['com.barebones.bbedit'],
  PhpStorm: ['com.jetbrains.PhpStorm'],
  PyCharm: ['com.jetbrains.PyCharm'],
  RubyMine: ['com.jetbrains.RubyMine'],
  TextMate: ['com.macromates.TextMate'],
  Brackets: ['io.brackets.appshell'],
  WebStorm: ['com.jetbrains.WebStorm'],
  Typora: ['abnerworks.Typora'],
  CodeRunner: ['com.krill.CodeRunner'],
  SlickEdit: [
    'com.slickedit.SlickEditPro2018',
    'com.slickedit.SlickEditPro2017',
    'com.slickedit.SlickEditPro2016',
    'com.slickedit.SlickEditPro2015',
  ],
  IntelliJ: ['com.jetbrains.intellij'],
  Xcode: ['com.apple.dt.Xcode'],
  GoLand: ['com.jetbrains.goland'],
  'Android Studio': ['com.google.android.studio'],
  Rider: ['com.jetbrains.rider'],
  Nova: ['com.panic.Nova'],
}
```

Can't get much more compact than that eh? Totally up to you. Then you can use `Object.entries` to iterate over them

```ts
  for (const [editor, bundleIdentifiers] of Object.entries(editors)) {
```

sergiou87 replied:
That's a good one! 🤩 

However I'm tempted to leave the current approach I have 😳 , since I think it makes the code (and the documentation in `editor-integration.md`) more consistent with Linux and Windows. What do you think?

niik replied:
Fine by me!

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,458 +10,279 @@ import {
+  {
+    name: 'JetBrains Webstorm',
+    registryKeys: [
+      // Webstorm 2018.3
```
Given that the uninstall key match the comment exactly I don't feel like these comments add much value

```suggestion
```

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,458 +10,279 @@ import {
+  {
+    name: 'JetBrains Phpstorm',
+    registryKeys: [
+      // PhpStorm 2019.2
```
Similar to Webstorm. Let's just drop these comments here

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -10,458 +10,279 @@ import {
+  {
+    name: 'JetBrains Rider',
+    registryKeys: [
+      // Rider 2019.3.4
```
It's already in the subKey
```suggestion
```

niik commented on app/src/lib/editors/win32.ts:
```diff
@@ -471,381 +292,65 @@ function getKeyOrEmpty(
+  const publisher = getKeyOrEmpty(keys, 'Publisher')
+  const installLocation = getKeyOrEmpty(
+    keys,
+    editor.installLocationRegistryKey ?? 'InstallLocation'
```
😍

niik reviewed (changes requested):
![image](https://user-images.githubusercontent.com/634063/105055530-74aa7880-5a73-11eb-9b22-123eb4059da2.png)

😻 this is such a wonderful cleanup @sergiou87! I had just a few tiny suggestions to give but I'm happy to ship this as is if you don't agree with them!

sergiou87 committed:
Drop redundant comments for Windows editors

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

niik reviewed (approved):
Feels so good to have all of this cleaned up. It's been collecting dust for much to long now!
